packages/insomnia-app/app/models/helpers/__tests__/settings.test.ts

```diff
@@ -0,0 +1,208 @@
+    getConfigSettings.mockReturnValue({ enableAnalytics: false });
+    const settings: Settings = {
+      ...models.settings.init(),
+      incognitoMode: false, // ensures incognito mode isn't affecting this test
+      enableAnalytics: true, // this intionally conflicts with the config
+    };
+
+    const controlledStatus = getControlledStatus(settings)('enableAnalytics');
+
+    expect(controlledStatus).toStrictEqual({
+      isControlled: true,
+      controller: 'insomnia-config',
+      value: false,
+    });
+  });
+
+  it('resolve conflicting value in the config and a controlling setting', () => {
+    getConfigSettings.mockReturnValue({ enableAnalytics: true }); // intentionally conflicts with incognito mode
+    const settings: Settings = {
+      ...models.settings.init(),
+      incognitoMode: true, // this intionally conflicts with the config
+      enableAnalytics: false, // this intionally conflicts with the config
+    };
+
+    const controlledStatus = getControlledStatus(settings)('enableAnalytics');
+
+    expect(controlledStatus).toStrictEqual({
+      isControlled: true,
+      controller: 'insomnia-config',
+      value: true,
+    });
+  });
+});
+
+describe('omitControlledSettings', () => {
+  it('omits config controlled settings', () => {
+    getConfigSettings.mockReturnValue({ disablePaidFeatureAds: true });
+    const settings = models.settings.init();
+
+    const result = omitControlledSettings(settings)({ disablePaidFeatureAds: false });
+
+    expect(result).toMatchObject({});
+  });
+
+  it('does not omit settings not controlled by the config', () => {
+    getConfigSettings.mockReturnValue({});
+    const settings = models.settings.init();
+
+    const result = omitControlledSettings(settings)({ disablePaidFeatureAds: true });
+
+    expect(result).toMatchObject({ disablePaidFeatureAds: true });
+  });
+
+  it('omits settings controlled by other settings', () => {
+    getConfigSettings.mockReturnValue({});
+    const settings: Settings = {
+      ...models.settings.init(),
+      incognitoMode: true,
+    };
+    const result = omitControlledSettings(settings)({ enableAnalytics: true });
+
+    expect(result).toMatchObject({});
+  });
+
+  it('does not omit settings not controlled by other settings', () => {
+    getConfigSettings.mockReturnValue({});
+    const settings = models.settings.init();
+    const result = omitControlledSettings(settings)({ disablePaidFeatureAds: true });
+
+    expect(result).toMatchObject({ disablePaidFeatureAds: true });
+  });
+});
+
+describe('getControlledSettings', () => {
+  it('overwrites config controlled settings', () => {
+    getConfigSettings.mockReturnValue({ disablePaidFeatureAds: true });
+    const settings: Settings = {
+      ...models.settings.init(),
+      disablePaidFeatureAds: false,
+    };
+
+    const result = getControlledSettings(settings);
+
+    expect(result).toMatchObject({ disablePaidFeatureAds: true });
```

Comment:
This should be checking for the entire settings object too, no? Same goes for the other tests in this describe block.

```suggestion
    expect(result).toMatchObject({ ...settings, disablePaidFeatureAds: true });
```

Reply:
it's done this way for separation of concerns.  This test is focused on only one small thing.  Therefore, it does not check whether the rest of the settings object is there.

I'm open to hearing more of your perspective on this.  Please say more about why these tests should be making assertions about things that are not directly related to the thing the test is testing.  Would another test that tests this thing (separately) be an acceptable solution?

Reply:
Because that function should only be touching controlled settings and leaving uncontrolled ones untouched. The assertion should therefore validate the negative case too, that is any properties it should not be touching remain untouched, and _only_ `disablePaidFeatureAds` is changed.

Reply:
the principle I'm suggesting we try to adhere to is that a unit test, almost by definition alone, should only test one thing. I gave the reason of "separation of concerns" but then your response lists two concerns (emphasis added)

> that function should only be touching controlled settings **and** leaving uncontrolled ones untouched.

These are different things, as far as I can tell.  We already have tests to ensure that it doesn't touch other settings, for example

- "does not omit settings not controlled by the config"
- "does not omit settings not controlled by other settings"
- "does not overwrite settings not controlled by the config"
- "does not overwrite settings not controlled by other settings"

I'm happy to hear any suggestions on how these tests can be improved, if you have any thoughts. :smile: 

___

Consider the following application of what you're suggesting:

<img height="200" src="https://user-images.githubusercontent.com/15232461/136969721-adf41a6a-cf76-447d-b001-55490b3b17ff.png" />

One might read this and ask "where did `allowNotificationRequests` come from??".  It's not clear why this is in the test assertion.  It's there because the test won't pass otherwise due to happenstance (i.e. the incognito mode business rule).  I would describe the above as a frail test.  I'm not trying to construct a strawman, you said

> Same goes for the other tests in this describe block.

So please let me know if this isn't one of the ones you talked about (it _is_ in this describe block).

____

I spent some time thinking about what you're suggesting and trying it out.  I think there's one test where the point you're making definitely is appropriate (because it's the thing the test is directly asserting about), and I've made that change in 2b5d9c671919f113128c4721e311873467474e02.  I'm happy to report that your suggestion lead me to realize that the test could have _more simply_ tested what it was aiming at by just removing the modification to user settings altogether.   Cool!

Beyond that, I added what you asked for in 3975a90c0071724db4171309057c6eb96917dd6c (and reverted it in the next commit) so that you can take a look at the full implications of the change.

Reply:
LGTM, the additional tests cover for it well